File: cost_tracker.py
"""
cost_tracker.py — Single source of truth for cost tracking in OpenClaw.

Extracted from gateway.py (was inline) and autonomous_runner.py (was duplicated).
All cost logging, calculation, and summary functions live here.

Primary backend: Supabase (real-time, queryable, multi-device)
Fallback: JSONL file (if Supabase is unreachable)
"""

import json
import os
import time
import logging
from datetime import datetime, timezone

logger = logging.getLogger("cost_tracker")

# ---------------------------------------------------------------------------
# Pricing table (per million tokens, USD)
# ---------------------------------------------------------------------------
COST_PRICING = {
    "claude-haiku-4-5-20251001":  {"input": 0.8,  "output": 4.0},
    "claude-sonnet-4-20250514":   {"input": 3.0,  "output": 15.0},
    "claude-opus-4-6":            {"input": 15.0, "output": 75.0},
    "claude-3-5-haiku-20241022":  {"input": 0.8,  "output": 4.0},
    "claude-3-5-sonnet-20241022": {"input": 3.0,  "output": 15.0},
    "kimi-2.5":                   {"input": 0.14, "output": 0.28},
    "kimi":                       {"input": 0.27, "output": 0.68},
    "m2.5":                       {"input": 0.30, "output": 1.20},
    "gemini-2.5-flash-lite":      {"input": 0.10, "output": 0.40},
    "gemini-2.5-flash":           {"input": 0.30, "output": 2.50},
    "gemini-3-flash-preview":     {"input": 0.00, "output": 0.00},
    "opencode":                    {"input": 0.05, "output": 0.08},
    "grok-3":                      {"input": 3.00, "output": 15.00},
    "grok-3-mini":                 {"input": 0.30, "output": 0.50},
    "grok-code-fast-1":            {"input": 0.30, "output": 0.50},
}

# Default fallback when a model is not in the table
_DEFAULT_PRICING = {"input": 3.0, "output": 15.0}


# ---------------------------------------------------------------------------
# Core helpers
# ---------------------------------------------------------------------------

def _calc_cost(model: str, tokens_in: int, tokens_out: int) -> float:
    """Return USD cost for the given token counts.
    Handles edge cases like zero or negative tokens to prevent invalid cost calculation.
    """
    if tokens_in < 0 or tokens_out < 0:
        logger.warning(f"Negative token counts provided for model {model}: tokens_in={tokens_in}, tokens_out={tokens_out}. Returning 0 cost.")
        return 0.0
    if tokens_in == 0 and tokens_out == 0:
        return 0.0

    pricing = COST_PRICING.get(model, _DEFAULT_PRICING)
    return round(
        (tokens_in * pricing["input"] + tokens_out * pricing["output"]) / 1_000_000,
        6,
    )


def get_cost_log_path() -> str:
    return os.environ.get(
        "OPENCLAW_COSTS_PATH",
        os.path.join(
            os.environ.get("OPENCLAW_DATA_DIR", "./data"),
            "costs",
            "costs.jsonl",
        ),
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def calculate_cost(model: str, tokens_input: int, tokens_output: int) -> float:
    """Calculate cost without logging."""
    return _calc_cost(model, tokens_input, tokens_output)


def _sb():
    """Lazy import supabase_client to avoid circular imports."""
    try:
        from supabase_client import table_insert, table_select, is_connected
        return {"insert": table_insert, "select": table_select, "connected": is_connected}
    except Exception:
        return None


def _use_supabase() -> bool:
    try:
        sb = _sb()
        return sb is not None and sb["connected"]()
    except Exception:
        return False


def log_cost_event(
    project: str = "openclaw",
    agent: str = "unknown",
    model: str = "unknown",
    tokens_input: int = 0,
    tokens_output: int = 0,
    cost: float = None,
    event_type: str = "api_call",
    metadata: dict = None,
    job_id: str = None,
    **kwargs,          # absorb unknown kwargs for compat
) -> float:
    """Calculate (or accept) cost, write to Supabase + JSONL fallback, return cost."""
    calculated_cost = cost if cost is not None else _calc_cost(model, tokens_input, tokens_output)

    # Try Supabase first
    if _use_supabase():
        sb = _sb()
        row = {
            "project": project,
            "agent": agent,
            "model": model,
            "tokens_input": tokens_input,
            "tokens_output": tokens_output,
            "cost_usd": calculated_cost,
            "created_at": datetime.now(timezone.utc).isoformat(),
        }
        if job_id:
            row["job_id"] = job_id
        result = sb["insert"]("costs", row)
        if result:
            logger.debug(f"Cost logged (Supabase): ${calculated_cost:.6f}")
            return calculated_cost
        logger.warning("Supabase cost insert failed, falling back to JSONL")

    # JSONL fallback
    entry = {
        "timestamp": time.time(),
        "type": event_type,
        "project": project,
        "agent": agent,
        "model": model,
        "tokens_in": tokens_input,
        "tokens_out": tokens_output,
        "cost": calculated_cost,
        "metadata": metadata or {},
    }
    if job_id:
        entry["job_id"] = job_id
    cost_path = get_cost_log_path()
    try:
        os.makedirs(os.path.dirname(cost_path), exist_ok=True)
        with open(cost_path, "a") as f:
            f.write(json.dumps(entry) + "\n")
    except Exception:
        pass
    return calculated_cost


def get_cost_metrics(days: int = None) -> dict:
    """Read costs and return aggregated metrics. Supabase-first, JSONL fallback."""
    entries = []

    if _use_supabase():
        sb = _sb()
        query = "order=created_at.desc"
        if days:
            cutoff = datetime.now(timezone.utc).isoformat()
            # PostgREST date filter
            from datetime import timedelta
            cutoff = (datetime.now(timezone.utc) - timedelta(days=days)).isoformat()
            query = f"created_at=gte.{cutoff}&{query}"
        rows = sb["select"]("costs", query, limit=5000)
        if rows is not None:
            for r in rows:
                entries.append({
                    "cost": float(r.get("cost_usd", 0)),
                    "agent": r.get("agent", "unknown"),
                    "project": r.get("project", "unknown"),
                    "model": r.get("model", "unknown"),
                })

    if not entries:
        # JSONL fallback
        cost_path = get_cost_log_path()
        try:
            with open(cost_path, "r") as f:
                for line in f:
                    line = line.strip()
                    if line:
                        try:
                            entries.append(json.loads(line))
                        except Exception:
                            pass
        except FileNotFoundError:
            pass

    total = sum(e.get("cost", e.get("cost_usd", 0)) for e in entries)
    by_agent: dict = {}
    for e in entries:
        a = e.get("agent", "unknown")
        by_agent[a] = by_agent.get(a, 0) + e.get("cost", e.get("cost_usd", 0))

    return {
        "total_cost":    round(total, 6),
        "entries_count": len(entries),
        "by_agent":      {k: round(v, 6) for k, v in by_agent.items()},
        "daily_total":   round(total, 6),
        "monthly_total": round(total, 6),
        "today_usd":     round(total, 6),
        "month_usd":     round(total, 6),
    }


def get_cost_summary() -> str:
    """One-liner summary string for dashboard/logs."""
    m = get_cost_metrics()
    return f"Total cost: ${m['total_cost']:.4f} across {m['entries_count']} API calls"

File: src/gateway/cost-tracker.ts
/**
 * Cost Tracker Module
 * Tracks API costs, logs to JSONL, and aggregates metrics
 * Supports Claude API pricing (Haiku, Sonnet, Opus)
 */

import * as fs from "fs";
import * as path from "path";

/**
 * Cost entry logged for each API call
 */
export interface CostTracker {
  project: string;
  agent: string;
  model: string;
  tokens_input: number;
  tokens_output: number;
  cost: number;
  timestamp: string;
}

/**
 * Aggregated cost metrics
 */
export interface CostMetrics {
  total_cost: number;
  by_project: Record<string, number>;
  by_model: Record<string, number>;
  by_agent: Record<string, number>;
  entries_count: number;
  timestamp_range: {
    first: string;
    last: string;
  };
}

/**
 * Pricing constants (Feb 2026 Claude API rates)
 * Input cost per million tokens, Output cost per million tokens
 */
const PRICING = {
  "claude-3-5-haiku-20241022": {
    input: 0.8, // $0.80 per million input tokens
    output: 4.0, // $4.00 per million output tokens
  },
  "claude-3-5-sonnet-20241022": {
    input: 3.0, // $3.00 per million input tokens
    output: 15.0, // $15.00 per million output tokens
  },
  "claude-opus-4-6": {
    input: 15.0, // $15.00 per million input tokens
    output: 75.0, // $75.00 per million output tokens
  },
  // Aliases for common references
  haiku: { input: 0.8, output: 4.0 },
  sonnet: { input: 3.0, output: 15.0 },
  opus: { input: 15.0, output: 75.0 },
};

/**
 * Get pricing for a model
 */
export function getPricing(model: string): { input: number; output: number } | null {
  return PRICING[model as keyof typeof PRICING] || null;
}

/**
 * Calculate cost for tokens
 */
export function calculateCost(model: string, tokens_input: number, tokens_output: number): number {
  const pricing = getPricing(model);
  if (!pricing) {
    console.warn(`Unknown model: ${model}, defaulting to Sonnet pricing`);
    return (tokens_input * PRICING.sonnet.input + tokens_output * PRICING.sonnet.output) / 1000000;
  }

  return (tokens_input * pricing.input + tokens_output * pricing.output) / 1000000;
}

/**
 * Get default cost log file path
 */
function getCostLogPath(): string {
  return process.env.OPENCLAW_COST_LOG || "/tmp/openclaw_costs.jsonl";
}

/**
 * Log a cost event to JSONL file (append-only)
 */
export async function logCostEvent(event: CostTracker): Promise<void> {
  const logPath = getCostLogPath();

  try {
    // Ensure directory exists
    const dir = path.dirname(logPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    // Append to JSONL (newline-delimited JSON)
    const line = JSON.stringify(event) + "\n";
    fs.appendFileSync(logPath, line, "utf-8");
  } catch (error) {
    console.error(`Failed to log cost event: ${error}`);
  }
}

/**
 * Read and parse JSONL cost log
 */
function readCostLog(): CostTracker[] {
  const logPath = getCostLogPath();

  if (!fs.existsSync(logPath)) {
    return [];
  }

  try {
    const content = fs.readFileSync(logPath, "utf-8");
    const lines = content.split("\n").filter((line) => line.trim());

    return lines
      .map((line) => {
        try {
          return JSON.parse(line) as CostTracker;
        } catch {
          console.warn(`Failed to parse cost log line: ${line}`);
          return null;
        }
      })
      .filter((entry) => entry !== null) as CostTracker[];
  } catch (error) {
    console.error(`Failed to read cost log: ${error}`);
    return [];
  }
}

/**
 * Get cost metrics aggregated from log
 */
export function getCostMetrics(timeWindow?: string): CostMetrics {
  const entries = readCostLog();

  if (entries.length === 0) {
    return {
      total_cost: 0,
      by_project: {},
      by_model: {},
      by_agent: {},
      entries_count: 0,
      timestamp_range: { first: "", last: "" },
    };
  }

  // Filter by time window if provided (e.g., "24h", "7d", "30d")
  let filtered = entries;
  if (timeWindow) {
    const now = new Date();
    const windowMs = parseTimeWindow(timeWindow);
    const cutoff = new Date(now.getTime() - windowMs);

    filtered = entries.filter((e) => new Date(e.timestamp) >= cutoff);
  }

  // Aggregate metrics
  const metrics: CostMetrics = {
    total_cost: 0,
    by_project: {},
    by_model: {},
    by_agent: {},
    entries_count: filtered.length,
    timestamp_range: {
      first: filtered.length > 0 ? filtered[0].timestamp : "",
      last: filtered.length > 0 ? filtered[filtered.length - 1].timestamp : "",
    },
  };

  for (const entry of filtered) {
    metrics.total_cost += entry.cost;
    metrics.by_project[entry.project] = (metrics.by_project[entry.project] || 0) + entry.cost;
    metrics.by_model[entry.model] = (metrics.by_model[entry.model] || 0) + entry.cost;
    metrics.by_agent[entry.agent] = (metrics.by_agent[entry.agent] || 0) + entry.cost;
  }

  return metrics;
}

/**
 * Parse time window string to milliseconds
 */
function parseTimeWindow(window: string): number {
  const match = window.match(/^(\d+)([hdm])$/);
  if (!match) {
    console.warn(`Invalid time window: ${window}, defaulting to 24h`);
    return 24 * 60 * 60 * 1000;
  }

  const value = parseInt(match[1], 10);
  const unit = match[2];

  switch (unit) {
    case "h":
      return value * 60 * 60 * 1000;
    case "d":
      return value * 24 * 60 * 60 * 1000;
    case "m":
      return value * 60 * 1000;
    default:
      return 24 * 60 * 60 * 1000;
  }
}

/**
 * Clear all cost data (for testing)
 */
export async function clearCostLog(): Promise<void> {
  const logPath = getCostLogPath();

  try {
    if (fs.existsSync(logPath)) {
      fs.unlinkSync(logPath);
      console.log("Cost log cleared");
    }
  } catch (error) {
    console.error(`Failed to clear cost log: ${error}`);
  }
}

/**
 * Get cost summary for debugging
 */
export function getCostSummary(): string {
  const metrics = getCostMetrics();
  const summary: string[] = [
    "=== Cost Summary ===",
    `Total Cost: $${metrics.total_cost.toFixed(4)}`,
    `Entries: ${metrics.entries_count}`,
    "",
    "By Project:",
    ...Object.entries(metrics.by_project).map(
      ([project, cost]) => `  ${project}: $${cost.toFixed(4)}`,
    ),
    "",
    "By Model:",
    ...Object.entries(metrics.by_model).map(([model, cost]) => `  ${model}: $${cost.toFixed(4)}`),
    "",
    "By Agent:",
    ...Object.entries(metrics.by_agent).map(([agent, cost]) => `  ${agent}: $${cost.toFixed(4)}`),
  ];

  return summary.join("\n");
}
